File: src/check.mjs
import fs from "node:fs";
import path from "node:path";
import { parseConcept, validateConcept } from "./concept.mjs";
import { renderIndex } from "./index-file.mjs";
import { TYPE_DIRS } from "./types.mjs";
export function check(bundle) {
  const problems = [];
  for (const dir of bundle.dirs()) {
    // Same set writeIndex uses, so "expected" here is exactly what `memory index`
    // would write; a placement violation still lands in entries (listConcepts
    // only parses) so it doesn't spuriously mark the index stale.
    const entries = bundle.listConcepts(dir);
    entries.dirRel = dir.rel;
    for (const typeDir of Object.values(TYPE_DIRS)) {
      const abs = path.join(dir.abs, typeDir); if (!fs.existsSync(abs)) continue;
      for (const name of fs.readdirSync(abs).filter((n) => n.endsWith(".md"))) {
        const rel = path.posix.join(dir.rel, typeDir, name);
        try { parseConcept(bundle.read(rel)); }
        catch (e) { problems.push({ rel, problem: e.message }); }
      }
    }
    for (const { rel, concept } of entries) {
      try { validateConcept(concept, { isRoot: dir.isRoot }); }
      catch (e) { problems.push({ rel, problem: e.message }); }
    }
    const indexRel = path.posix.join(dir.rel, "index.md");
    const expected = renderIndex(entries, { isRoot: dir.isRoot });
    const actual = bundle.read(indexRel);
    if (actual !== expected) problems.push({ rel: indexRel, problem: actual == null ? "missing index" : "index differs from regeneration" });
  }
  return { ok: problems.length === 0, problems };
}

File: memory-go/check.go
package memory

import (
	"os"
	"path"
	"path/filepath"
	"slices"
	"strings"
)

type Problem struct {
	Rel     string `json:"rel"`
	Problem string `json:"problem"`
}

type CheckResult struct {
	OK       bool      `json:"ok"`
	Problems []Problem `json:"problems"`
}

var checkSkipDirs = map[string]bool{".git": true, ".state": true, ".locks": true, "node_modules": true}

// Check is check(): every stray .md outside a type directory is reported and
// scanned for secrets, every concept must parse and sit in a legal
// directory, and every index must equal its regeneration.
func Check(b *Bundle) (CheckResult, error) {
	var problems []Problem
	if err := walkStray(b, b.Root, "", &problems); err != nil {
		return CheckResult{}, err
	}
	dirs, err := b.Dirs()
	if err != nil {
		return CheckResult{}, err
	}
	for _, dir := range dirs {
		entries, err := b.ListConcepts(dir)
		if err != nil {
			return CheckResult{}, err
		}
		for _, typeDir := range TypeDirOrder {
			files, err := os.ReadDir(filepath.Join(dir.Abs, typeDir))
			if err != nil {
				continue
			}
			for _, f := range files {
				if !strings.HasSuffix(f.Name(), ".md") {
					continue
				}
				rel := path.Join(dir.Rel, typeDir, f.Name())
				text, _ := b.Read(rel)
				if _, err := ParseConcept(text); err != nil {
					problems = append(problems, Problem{rel, err.Error()})
				}
			}
		}
		for _, e := range entries {
			if err := ValidateConcept(e.Concept, dir.IsRoot); err != nil {
				problems = append(problems, Problem{e.Rel, err.Error()})
			}
		}
		indexRel := joinRel(dir.Rel, "index.md")
		expected := RenderIndex(entries, dir.Rel, dir.IsRoot)
		actual, ok := b.Read(indexRel)
		switch {
		case !ok:
			problems = append(problems, Problem{indexRel, "missing index"})
		case actual != expected:
			problems = append(problems, Problem{indexRel, "index differs from regeneration"})
		}
	}
	return CheckResult{OK: len(problems) == 0, Problems: problems}, nil
}

func walkStray(b *Bundle, abs, rel string, problems *[]Problem) error {
	entries, err := os.ReadDir(abs)
	if err != nil {
		return err
	}
	for _, e := range entries {
		if e.IsDir() {
			if checkSkipDirs[e.Name()] {
				continue
			}
			if err := walkStray(b, filepath.Join(abs, e.Name()), path.Join(rel, e.Name()), problems); err != nil {
				return err
			}
			continue
		}
		if !strings.HasSuffix(e.Name(), ".md") || e.Name() == "index.md" || e.Name() == "log.md" {
			continue
		}
		if slices.Contains(TypeDirOrder, path.Base(rel)) {
			continue
		}
		fileRel := path.Join(rel, e.Name())
		*problems = append(*problems, Problem{fileRel, "unrecognized file"})
		text, _ := b.Read(fileRel)
		if hit := FindSecret(text); hit != nil {
			*problems = append(*problems, Problem{fileRel, "secret (" + hit.Name + ")"})
		}
	}
	return nil
}
